content/Projects/Observability/Platform/faq.md

```diff
@@ -0,0 +1,138 @@
+
+1. Ensure the `serviceMonitorSelector` in the `Prometheus` CR matches the key in the `ServiceMonitor` labels.
+2. The `Service` you want to scrape *must* have an explicitly named port.
+3. The `ServiceMonitor` *must* reference the `port` by this name.
+
+Assuming this criteria is met but the metrics don't exist, we can try debug the following.
+
+```bash
+oc -n openshift-monitoring get secret prometheus-k8s -o json | jq -r '.data["prometheus.yaml.gz"]' | base64 -d | gunzip | grep "<service-monitor-name>"
+```
+
+If the `ServiceMonitor` does not exist in the output, the next step would be to investigate the logs of both `prometheus` and the `prometheus-operator` for errors.
+
+Assuming it does exist then we know `prometheus-operator` is doing its job. Double check the `ServiceMonitor` definition.
+
+Check the service discovery endpoint to ensure Prometheus can discover the target. It will need the appropriate RBAC to do so. An example can be found [here](https://github.com/openshift/cluster-monitoring-operator/blob/23201e012586d4864ca23593621f843179c47412/assets/prometheus-k8s/role-specific-namespaces.yaml#L35-L50).
+
+There is a possibility Prometheus has not loaded the configuration yet. The following metrics will help to determine if that is in fact the case or if there are errors in the configuration:
+
+```bash
+prometheus_config_last_reload_success_timestamp_seconds
+prometheus_config_last_reload_successful
+```
+
+If the output determines that the configuration should by now have been applied, proceed again with examining the logs for errors.
+
+### TargetDown alert firing
+
+We have, in the past seen cases where the `TargetDown` alert was firing when all endpoints appeared to be up. The following commands fetch some useful metrics to help identify the cause.
+
+As the alert fires, get the list of active targets in Prometheus
+
+```bash
+oc exec -n openshift-monitoring prometheus-k8s-0 -c prometheus -- curl http://localhost:9090/api/v1/targets?state=active > targets.prometheus-k8s-0.json
+
+oc exec -n openshift-monitoring prometheus-k8s-1 -c prometheus -- curl http://localhost:9090/api/v1/targets?state=active > targets.prometheus-k8s-1.json
+```
+
+---
+
+Reports all targets that Prometheus couldn’t connect to with some reason (timeout, refused, …)
+
+A `dialer_name` can be passed as a label to limit the query to interesting components. For example `{dialer_name=~".+openshift-.*"}`.
+
+```bash
+oc exec -n openshift-monitoring prometheus-k8s-0 -c prometheus -- curl http://localhost:9090/api/v1/query --data-urlencode 'query=net_conntrack_dialer_conn_failed_total{} > 1' > net_conntrack_dialer_conn_failed_total.prometheus-k8s-0.json
```

Comment:
(nit) we could use rate() here to exclude targets that couldn't be scraped  in the past but are ok now.

```suggestion
oc exec -n openshift-monitoring prometheus-k8s-0 -c prometheus -- curl http://localhost:9090/api/v1/query --data-urlencode 'query=rate(net_conntrack_dialer_conn_failed_total{}[1h]) > 0' > net_conntrack_dialer_conn_failed_total.prometheus-k8s-0.json
```